lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3264,7 +3267,9 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
       return castKind;
 
     case CheckedCastKind::Unresolved:
-      return failed();
+      if (contextKind == CheckedCastContextKind::Coercion)
+        return failed();
+      return castKind;
```

Comment:
Could you add a comment to explain this logic? Something along the lines of:

```cxx
Even if we know the elements cannot be downcast, we cannot return failed() here as
it's possible for an empty Array, Set or Dictionary to be cast to any element type.
The one exception to this is when we're checking whether we can treat a coercion as
a checked cast – we don't want to tell the user to use as!, as it's probably the
wrong suggestion.
```

Reply:
Sure! I think this explains everything.
Thanks, @hamishknight  :)

Reply:
Fixed!